Option based pricing

thenchev opened the pull request.

thenchev committed:
Option based pricing

shabananavas commented on modules/commerce_option_pricing/commerce_option_pricing.install:
```diff
@@ -0,0 +1,17 @@
+ */
+function commerce_option_pricing_install() {
+  $entity_type = \Drupal::service('entity_type.manager')->getDefinition('commerce_option_option_value');
+  \Drupal::service('entity.definition_update_manager')->updateEntityType($entity_type);
```
Why are we updating the entity at this point? We haven't made any changes to the entity yet right?

shabananavas replied:
Ah, it's because of this right: `commerce_option_pricing_entity_base_field_info()`

shabananavas commented on modules/commerce_option_pricing/src/Event/OptionPriceResolverEvent.php:
```diff
@@ -0,0 +1,110 @@
+
+namespace Drupal\commerce_option_pricing\Event;
+
+use Drupal\commerce\PurchasableEntityInterface;
```
Grouping issue here.

shabananavas commented on modules/commerce_option_pricing/src/Resolver/OptionPriceResolver.php:
```diff
@@ -0,0 +1,110 @@
+
+use Drupal\Core\Entity\EntityTypeManagerInterface;
+use Drupal\Core\TempStore\PrivateTempStoreFactory;
+use Symfony\Component\EventDispatcher\EventDispatcherInterface;
```
Grouping

shabananavas commented on modules/commerce_option_pricing/src/Resolver/OptionPriceResolver.php:
```diff
@@ -0,0 +1,110 @@
+      $addedPrice = $addedPrice + $option->pricing->first()->toPrice()->getNumber();
+    }
+
+    $field_name = $context->getData('field_name', 'price');
```
Could we add some comments in this function to describe the different scenarios so that it provides more clarity for why we're doing if/else etc.

shabananavas commented on modules/commerce_option_pricing/src/Resolver/OptionPriceResolver.php:
```diff
@@ -0,0 +1,110 @@
+
+      return new Price($totalPrice, $currencyCode);
+    }
+    elseif ($entity->hasField($field_name) && !$entity->get($field_name)->isEmpty()) {
```
What cases does this else if cover?

shabananavas replied:
@thenchev As I'm looking at it, couldn't we rewrite this whole if/else like this, starting from line 82:

```php
  $field_name = $context->getData('field_name', 'price');
  if (!$entity->hasField($field_name) || $entity->get($field_name)->isEmpty()) {
    return NULL;
  }

  // Handle the base case where we use the default price field.
  if ($field_name == 'price') {
    $base_price = $entity->getPrice()->getNumber();
    $currencyCode = $entity->getPrice()->getCurrencyCode();
  }
  // Handle the case where the field is not the default price field and we get
  // the price from another price field.
  else {
    $base_price = $entity->get($field_name)->first()->toPrice()->getNumber();
    $currencyCode = $entity->get($field_name)->first()->toPrice()->getCurrencyCode();
  }

  $totalPrice = Calculator::multiply(Calculator::add($base_price, $addedPrice), $quantity);
  $event = new OptionPriceResolverEvent($entity, $quantity, $options, $totalPrice);

  // Allow others to override this price.
  $this->eventDispatcher
    ->dispatch(OptionPricingEvents::OPTION_PRICING_RESOLVE, $event);
  $totalPrice = $event->getTotalPrice();

  return new Price($totalPrice, $currencyCode);
```

safalliajoseph replied:
@thenchev @shabananavas I do not understand why we need a event dispatcher here. Is it for allowing other modules to alter the price?

shabananavas replied:
Saf, yeah, I had asked Milos to add that in to allow other modules to alter the price. Because, perhaps they might have other custom implementations. They could always implement their own custom resolver, but it would still be a good option to give them.

safalliajoseph replied:
@shabananavas Yeah they can implement a custom resolver. So I have removed this event dispatcher now.

shabananavas commented:
@thenchev Added a couple of comments. Love that you're adding tests for every feature!

thenchev committed:
Address feedback

Commit:
#3115861 - Refactor the pricing calculation
Alter the addto cart form and pass options value via context to price resolver

safalliajoseph commented:
@thenchev @shabananavas I have modified the code. I removed tempstore and instead extended the Addtocart form and passed the selected options via context to Price resolver. Also used the price api methods for calculations. I tested this in accurate industries and seems to be working as expected.

@thenchev I now understand why you used temp storage for storing options. We do not get order item object in price resolver, hence cannot get the selected options. I used another method, extended the add to cart form, the submit function of addToCart form calls the price resolvers, so we can pass in options as context variable to the resolver. 

@shabananavas Could you please review the code.

shabananavas commented on modules/commerce_option_pricing/tests/src/Kernel/OptionPriceResolverTest.php:
```diff
@@ -0,0 +1,123 @@
+ *
+ * @group commerce_option_pricing
+ */
+class OptionPriceResolverTest extends CommerceKernelTestBase {
```
I think with the new changes, chances are, these tests might not be passing. If not, we can remove it from this MR and work on fixing it later as we are short on time.

safalliajoseph replied:
I have modified the test to follow the new code. So it should be working now.

shabananavas commented on src/EventSubscriber/CartEventSubscriber.php:
```diff
@@ -33,42 +33,11 @@ public function __construct(PrivateTempStoreFactory $tempstore_private) {
-   * @param \Symfony\Component\EventDispatcher\Event $event
-   *   The dispatched event.
-   */
-  public function addOptions(Event $event) {
```
@safjoseph So, with removing this, does merging of the cart items still continue to work depending on the selected options? Will the cart items will automatically increment the quantities (instead of adding a new item to the cart) if the user selected a product with a specific option and add them to cart, then, comes back and selects the same product with the same option again?

safalliajoseph replied:
Yep. It works as expected.

shabananavas commented on src/Form/AddToCartForm.php:
```diff
@@ -0,0 +1,80 @@
+      $store = $this->selectStore($purchased_entity);
+
+      // Pass in options via context to the price resolver.
+      $context = new Context(
```
Great work!

safalliajoseph replied:
Thank you. It took me a while to figure out this.

shabananavas commented:
@safjoseph Excellent work. Thanks for refactoring that. We're learning newer and newer things as we go 😄 

I just had a couple of questions, please take a look. Thanks.

Commit:
#3115861 - Alter order refresh service class and pass options to price resolver

safalliajoseph commented:
@shabananavas I have addressed the issues you raised. I tested this with accurate and the pricing seems to work as expected. Could you please review. Thanks!

thenchev commented:
> @thenchev @shabananavas I have modified the code. I removed tempstore and instead extended the Addtocart form and passed the selected options via context to Price resolver. Also used the price api methods for calculations. I tested this in accurate industries and seems to be working as expected.
> 
> @thenchev I now understand why you used temp storage for storing options. We do not get order item object in price resolver, hence cannot get the selected options. I used another method, extended the add to cart form, the submit function of addToCart form calls the price resolvers, so we can pass in options as context variable to the resolver.
> 
> @shabananavas Could you please review the code.

I knew there was something :)
Great idea with the context, I always forget it exists... maybe to abstract for me

shabananavas commented:
@safalliajoseph Looks good. Great job guys. Thanks!! Please post an MR on d.org.

safalliajoseph committed:
3115861 option based shipping weight (#8)

Add commerce option based shipping weight

Co-authored-by: Safallia Joseph <sjoseph@acromedia.com>
